Sources/AsyncHTTPClient/HTTPDecompressor.swift

```diff
@@ -0,0 +1,151 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2018-2019 Swift Server Working Group and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOExtrasZlib
+import NIO
+import NIOHTTP1
+
+private enum CompressionAlgorithm: String {
+    case gzip
+    case deflate
+}
+
+extension z_stream {
+    public mutating func inflatePart(input: inout ByteBuffer, allocator: ByteBufferAllocator, consumer: (ByteBuffer) -> Void) {
+        input.readWithUnsafeMutableReadableBytes { dataPtr in
+            let typedPtr = dataPtr.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            let typedDataPtr = UnsafeMutableBufferPointer(start: typedPtr, count: dataPtr.count)
+
+            self.avail_in = UInt32(typedDataPtr.count)
+            self.next_in = typedDataPtr.baseAddress!
+
+            repeat {
+                var buffer = allocator.buffer(capacity: 16384)
+                self.inflatePart(to: &buffer)
+                consumer(buffer)
+            } while self.avail_out == 0
+
+            return Int(self.avail_in)
+        }
```

Comment:
The original code has a defer block that you _must_ use here for the sake of correctness: specifically, you must unset all the pointers on the `z_stream` structure when you're done.

Reply:
Good catch, fixed, thanks!